> Introdução à Ciência dos Dados \
> Ciência da Computação – UFV/Florestal


* **Nome:** %s
* **Matrícula:**  %d

# Aula Prática 07
## Regras de Associação e o Algoritmo Apriori

**Objetivo:** Aplicar o algoritmo Apriori para extrair regras de associação de dados reais.
**Pré-requisitos:** Linguagem de programação Python, estatística básica \
**Meta:** Ao final da prática, o aluno será capaz de preparar dados e extrair regras de associação desses dados utilizado o algoritmo Apriori.

----

In [1]:
### NÃO REMOVA ESSA CÉLULA! ####
!pip --no-cache-dir install --upgrade learntools_dados_ufv

#!wget https://raw.githubusercontent.com/NESPEDUFV/CCF425-resources/master/p7/Market_Basket_Optimisation.csv

In [2]:
#### Não altere essa célula!!! ####
import pandas as pd
import string
import random
import numpy as np

from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

from learntools_dados_ufv.core import binder; binder.bind(globals())
from learntools_dados_ufv.ccf425.pratica_7b import *

# Roteiro

## Estruturando os dados

### Criar o conjunto de dados

In [3]:
dataset = [['Leite', 'Cebola', 'Nutela', 'Feijão', 'Ovo', 'Iogurte'],
          ['Arroz', 'Cebola', 'Nutela', 'Feijão', 'Ovo', 'Iogurte'],
          ['Leite', 'Maça', 'Feijão', 'Ovo'],
          ['Leite', 'Chocolate', 'Pipoca', 'Feijão', 'Iogurte'],
          ['Pipoca', 'Cebola', 'Feijão', 'Sorvete', 'Ovo']]

dataset

[['Leite', 'Cebola', 'Nutela', 'Feijão', 'Ovo', 'Iogurte'],
 ['Arroz', 'Cebola', 'Nutela', 'Feijão', 'Ovo', 'Iogurte'],
 ['Leite', 'Maça', 'Feijão', 'Ovo'],
 ['Leite', 'Chocolate', 'Pipoca', 'Feijão', 'Iogurte'],
 ['Pipoca', 'Cebola', 'Feijão', 'Sorvete', 'Ovo']]

### Transformar no formato para ser utilizado no Apriori

In [4]:
transactions_oht = TransactionEncoder()
transactions = transactions_oht.fit(dataset).transform(dataset)
transactions



array([[False,  True, False,  True,  True,  True, False,  True,  True,
        False, False],
       [ True,  True, False,  True,  True, False, False,  True,  True,
        False, False],
       [False, False, False,  True, False,  True,  True, False,  True,
        False, False],
       [False, False,  True,  True,  True,  True, False, False, False,
         True, False],
       [False,  True, False,  True, False, False, False, False,  True,
         True,  True]])

In [5]:
df = pd.DataFrame(transactions, columns=transactions_oht.columns_)
df

Arroz  Cebola  Chocolate  Feijão  Iogurte  Leite   Maça  Nutela    Ovo  \
0  False    True      False    True     True   True  False    True   True   
1   True    True      False    True     True  False  False    True   True   
2  False   False      False    True    False   True   True   False   True   
3  False   False       True    True     True   True  False   False  False   
4  False    True      False    True    False  False  False   False   True   

   Pipoca  Sorvete  
0   False    False  
1   False    False  
2   False    False  
3    True    False  
4    True     True

## Calcular conjunto de itens frequentes

Com suporte mínimo de 0.6

In [6]:
# Calcula os conjuntos de itens frequentes, com suporte minimo de 0.6
frequent_itemsets = apriori(df, min_support=0.6, use_colnames=True)
frequent_itemsets

support               itemsets
0       0.6               (Cebola)
1       1.0               (Feijão)
2       0.6              (Iogurte)
3       0.6                (Leite)
4       0.8                  (Ovo)
5       0.6       (Feijão, Cebola)
6       0.6          (Cebola, Ovo)
7       0.6      (Feijão, Iogurte)
8       0.6        (Feijão, Leite)
9       0.8          (Feijão, Ovo)
10      0.6  (Feijão, Cebola, Ovo)

## Calcular regras de associacão utilizando o Apriori

Com filtro de confiança de 0.1

In [7]:
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.1)
rules

antecedents       consequents  antecedent support  \
0           (Feijão)          (Cebola)                 1.0   
1           (Cebola)          (Feijão)                 0.6   
2           (Cebola)             (Ovo)                 0.6   
3              (Ovo)          (Cebola)                 0.8   
4           (Feijão)         (Iogurte)                 1.0   
5          (Iogurte)          (Feijão)                 0.6   
6           (Feijão)           (Leite)                 1.0   
7            (Leite)          (Feijão)                 0.6   
8           (Feijão)             (Ovo)                 1.0   
9              (Ovo)          (Feijão)                 0.8   
10  (Feijão, Cebola)             (Ovo)                 0.6   
11     (Feijão, Ovo)          (Cebola)                 0.8   
12     (Cebola, Ovo)          (Feijão)                 0.6   
13          (Feijão)     (Cebola, Ovo)                 1.0   
14          (Cebola)     (Feijão, Ovo)                 0.6   
15             (Ovo)  (Feijão, Cebola)                 0.8   

    consequent support  support  confidence  lift  leverage  conviction  \
0                  0.6      0.6        0.60  1.00      0.00         1.0   
1                  1.0      0.6        1.00  1.00      0.00         inf   
2                  0.8      0.6        1.00  1.25      0.12         inf   
3                  0.6      0.6        0.75  1.25      0.12         1.6   
4                  0.6      0.6        0.60  1.00      0.00         1.0   
5                  1.0      0.6        1.00  1.00      0.00         inf   
6                  0.6      0.6        0.60  1.00      0.00         1.0   
7                  1.0      0.6        1.00  1.00      0.00         inf   
8                  0.8      0.8        0.80  1.00      0.00         1.0   
9                  1.0      0.8        1.00  1.00      0.00         inf   
10                 0.8      0.6        1.00  1.25      0.12         inf   
11                 0.6      0.6        0.75  1.25      0.12         1.6   
12                 1.0      0.6        1.00  1.00      0.00         inf   
13                 0.6      0.6        0.60  1.00      0.00         1.0   
14                 0.8      0.6        1.00  1.25      0.12         inf   
15                 0.6      0.6        0.75  1.25      0.12         1.6   

    zhangs_metric  
0             0.0  
1             0.0  
2             0.5  
3             1.0  
4             0.0  
5             0.0  
6             0.0  
7             0.0  
8             0.0  
9             0.0  
10            0.5  
11            1.0  
12            0.0  
13            0.0  
14            0.5  
15            1.0

In [8]:
# Filtro por 'confidence'. Pode ser feito por outros campos também
rules[rules['confidence'] > 0.75]

antecedents    consequents  antecedent support  consequent support  \
1           (Cebola)       (Feijão)                 0.6                 1.0   
2           (Cebola)          (Ovo)                 0.6                 0.8   
5          (Iogurte)       (Feijão)                 0.6                 1.0   
7            (Leite)       (Feijão)                 0.6                 1.0   
8           (Feijão)          (Ovo)                 1.0                 0.8   
9              (Ovo)       (Feijão)                 0.8                 1.0   
10  (Feijão, Cebola)          (Ovo)                 0.6                 0.8   
12     (Cebola, Ovo)       (Feijão)                 0.6                 1.0   
14          (Cebola)  (Feijão, Ovo)                 0.6                 0.8   

    support  confidence  lift  leverage  conviction  zhangs_metric  
1       0.6         1.0  1.00      0.00         inf            0.0  
2       0.6         1.0  1.25      0.12         inf            0.5  
5       0.6         1.0  1.00      0.00         inf            0.0  
7       0.6         1.0  1.00      0.00         inf            0.0  
8       0.8         0.8  1.00      0.00         1.0            0.0  
9       0.8         1.0  1.00      0.00         inf            0.0  
10      0.6         1.0  1.25      0.12         inf            0.5  
12      0.6         1.0  1.00      0.00         inf            0.0  
14      0.6         1.0  1.25      0.12         inf            0.5

# Exercícios

Os exercícios abaixo são referentes ao conjunto de dados `Market_Basket_Optimisation.csv`. Esse conjunto possui compras considerando 20 produtos distintos.

Execute e não remova a célula abaixo antes de iniciar os exercícios. Faça todas as atividades manipulando a variável `df`.

In [9]:
# Atribui o dataframe com o estado "original" em uma variável isolada
original_df = pd.read_csv('./Market_Basket_Optimisation.csv',header=None)
df = original_df.copy()
df

0                  1            2                 3   \
0            shrimp            almonds      avocado    vegetables mix   
1           burgers          meatballs         eggs               NaN   
2           chutney                NaN          NaN               NaN   
3            turkey            avocado          NaN               NaN   
4     mineral water               milk   energy bar  whole wheat rice   
...             ...                ...          ...               ...   
7496         butter         light mayo  fresh bread               NaN   
7497        burgers  frozen vegetables         eggs      french fries   
7498        chicken                NaN          NaN               NaN   
7499       escalope          green tea          NaN               NaN   
7500           eggs    frozen smoothie  yogurt cake    low fat yogurt   

                4                 5     6               7             8   \
0     green grapes  whole weat flour  yams  cottage cheese  energy drink   
1              NaN               NaN   NaN             NaN           NaN   
2              NaN               NaN   NaN             NaN           NaN   
3              NaN               NaN   NaN             NaN           NaN   
4        green tea               NaN   NaN             NaN           NaN   
...            ...               ...   ...             ...           ...   
7496           NaN               NaN   NaN             NaN           NaN   
7497     magazines         green tea   NaN             NaN           NaN   
7498           NaN               NaN   NaN             NaN           NaN   
7499           NaN               NaN   NaN             NaN           NaN   
7500           NaN               NaN   NaN             NaN           NaN   

                9               10         11     12     13             14  \
0     tomato juice  low fat yogurt  green tea  honey  salad  mineral water   
1              NaN             NaN        NaN    NaN    NaN            NaN   
2              NaN             NaN        NaN    NaN    NaN            NaN   
3              NaN             NaN        NaN    NaN    NaN            NaN   
4              NaN             NaN        NaN    NaN    NaN            NaN   
...            ...             ...        ...    ...    ...            ...   
7496           NaN             NaN        NaN    NaN    NaN            NaN   
7497           NaN             NaN        NaN    NaN    NaN            NaN   
7498           NaN             NaN        NaN    NaN    NaN            NaN   
7499           NaN             NaN        NaN    NaN    NaN            NaN   
7500           NaN             NaN        NaN    NaN    NaN            NaN   

          15                 16               17       18         19  
0     salmon  antioxydant juice  frozen smoothie  spinach  olive oil  
1        NaN                NaN              NaN      NaN        NaN  
2        NaN                NaN              NaN      NaN        NaN  
3        NaN                NaN              NaN      NaN        NaN  
4        NaN                NaN              NaN      NaN        NaN  
...      ...                ...              ...      ...        ...  
7496     NaN                NaN              NaN      NaN        NaN  
7497     NaN                NaN              NaN      NaN        NaN  
7498     NaN                NaN              NaN      NaN        NaN  
7499     NaN                NaN              NaN      NaN        NaN  
7500     NaN                NaN              NaN      NaN        NaN  

[7501 rows x 20 columns]

In [10]:
len(df.columns)

20

## 1. Faça uma preparação dos dados.
Antes de fazer a análise de regras de associação, é necessário que você trate alguns dados desse dataset. Essa prática é **extremamente** importante e comum, já que os dados que encontramos ~~quase nunca~~ nem sempre estão prontos para serem manipulados.

### 1.a. Crie um dataframe com base nas transações, usando o TransactionEncoder e manipulação de Array

Chame a função `q1.a.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q1.a.hint()`

In [11]:
# Resposta da 1
df = df.fillna('') # Substitui os valores nulos por uma string vazia
#df = df.replace(to_replace =" asparagus", value ="asparagus") # Corrige um erro de digitação
transactions_array = []
for i in range(len(df)):
    transactions_array.append(df.iloc[i].values)
transactions_array

[array(['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes',
        'whole weat flour', 'yams', 'cottage cheese', 'energy drink',
        'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad',
        'mineral water', 'salmon', 'antioxydant juice', 'frozen smoothie',
        'spinach', 'olive oil'], dtype=object),
 array(['burgers', 'meatballs', 'eggs', '', '', '', '', '', '', '', '', '',
        '', '', '', '', '', '', '', ''], dtype=object),
 array(['chutney', '', '', '', '', '', '', '', '', '', '', '', '', '', '',
        '', '', '', '', ''], dtype=object),
 array(['turkey', 'avocado', '', '', '', '', '', '', '', '', '', '', '',
        '', '', '', '', '', '', ''], dtype=object),
 array(['mineral water', 'milk', 'energy bar', 'whole wheat rice',
        'green tea', '', '', '', '', '', '', '', '', '', '', '', '', '',
        '', ''], dtype=object),
 array(['low fat yogurt', '', '', '', '', '', '', '', '', '', '', '', '',
        '', '', '', '', '', '', ''], d

In [12]:
transactions_oht = TransactionEncoder()
transactions = transactions_oht.fit(transactions_array).transform(transactions_array)
transactions

array([[False, False,  True, ...,  True, False, False],
       [ True, False, False, ..., False, False, False],
       [ True, False, False, ..., False, False, False],
       ...,
       [ True, False, False, ..., False, False, False],
       [ True, False, False, ..., False, False, False],
       [ True, False, False, ..., False,  True, False]])

In [13]:
df = pd.DataFrame(transactions, columns=transactions_oht.columns_)
df = df.drop('', axis=1) #Exclusão da coluna vazia
df

q1.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



## 2. Faça o cálculo dos itemsets frequentes e regras de associação.
Utilize o algoritmo _apriori_.

### 2.a. A partir de `df`, calcule os itemsets frequentes utilizando `apriori`. 

Utilize o suporte mínimo de 0.01. Utilize o paramêtro `use_colnames=True`.

Armazene-a na variável `frequent_itemsets`.

Chame a função `q3.a.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q3.a.hint()`

In [14]:
# Resposta da 2.a

# q2.a.hint()

frequent_itemsets = apriori(df, min_support=0.01, use_colnames=True)
q2.a.check()
frequent_itemsets


<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



support                                 itemsets
0    0.020397                                (almonds)
1    0.033329                                (avocado)
2    0.010799                         (barbecue sauce)
3    0.014265                              (black tea)
4    0.011465                             (body spray)
..        ...                                      ...
252  0.011065       (mineral water, ground beef, milk)
253  0.017064  (mineral water, ground beef, spaghetti)
254  0.015731         (mineral water, spaghetti, milk)
255  0.010265    (mineral water, olive oil, spaghetti)
256  0.011465     (mineral water, pancakes, spaghetti)

[257 rows x 2 columns]

### 2.b. A partir de `frequent_itemsets`, calcule as regras de associação. 

Utilize filtro de confiança de `0.3`.

Armazene-a na variável `rules`.

Chame a função `q2.b.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q2.b.hint()`

In [15]:
# Resposta da 2.b

# q2.b.hint()

rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.3)
q2.b.check()
rules

<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



antecedents      consequents  antecedent support  \
0                    (avocado)  (mineral water)            0.033329   
1                    (burgers)           (eggs)            0.087188   
2                       (cake)  (mineral water)            0.081056   
3                    (cereals)  (mineral water)            0.025730   
4                    (chicken)  (mineral water)            0.059992   
..                         ...              ...                 ...   
58           (milk, spaghetti)  (mineral water)            0.035462   
59  (mineral water, olive oil)      (spaghetti)            0.027596   
60      (olive oil, spaghetti)  (mineral water)            0.022930   
61   (mineral water, pancakes)      (spaghetti)            0.033729   
62       (pancakes, spaghetti)  (mineral water)            0.025197   

    consequent support   support  confidence      lift  leverage  conviction  \
0             0.238368  0.011598    0.348000  1.459926  0.003654    1.168147   
1             0.179709  0.028796    0.330275  1.837830  0.013128    1.224818   
2             0.238368  0.027463    0.338816  1.421397  0.008142    1.151921   
3             0.238368  0.010265    0.398964  1.673729  0.004132    1.267198   
4             0.238368  0.022797    0.380000  1.594172  0.008497    1.228438   
..                 ...       ...         ...       ...       ...         ...   
58            0.238368  0.015731    0.443609  1.861024  0.007278    1.368879   
59            0.174110  0.010265    0.371981  2.136468  0.005460    1.315071   
60            0.238368  0.010265    0.447674  1.878079  0.004799    1.378954   
61            0.174110  0.011465    0.339921  1.952333  0.005593    1.251198   
62            0.238368  0.011465    0.455026  1.908923  0.005459    1.397557   

    zhangs_metric  
0        0.325896  
1        0.499424  
2        0.322617  
3        0.413162  
4        0.396502  
..            ...  
58       0.479672  
59       0.547034  
60       0.478514  
61       0.504819  
62       0.488452  

[63 rows x 10 columns]

## 3. Analise as regras obtidas.

### 3.a. A partir de `rules`, recupere as regras com suporte maior ou igual a 0.015 e confiança maior ou igual a 0.4. 

Armazene-a na variável `rules_a`.

Chame a função `q3.a.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q3.a.hint()`

In [16]:
# Resposta da 3.a

# q3.a.hint()

rules_a = rules[(rules['support'] >= 0.015) & (rules['confidence'] >= 0.4)]

q3.a.check()
rules_a

<IPython.core.display.Javascript object>

<span style="color:#33cc33">CORRETO:</span> 



antecedents      consequents  antecedent support  \
15                 (ground beef)  (mineral water)            0.098254   
23                   (olive oil)  (mineral water)            0.065858   
26                      (salmon)  (mineral water)            0.042528   
28                        (soup)  (mineral water)            0.050527   
44        (chocolate, spaghetti)  (mineral water)            0.039195   
55  (mineral water, ground beef)      (spaghetti)            0.040928   
56      (ground beef, spaghetti)  (mineral water)            0.039195   
58             (milk, spaghetti)  (mineral water)            0.035462   

    consequent support   support  confidence      lift  leverage  conviction  \
15            0.238368  0.040928    0.416554  1.747522  0.017507    1.305401   
23            0.238368  0.027596    0.419028  1.757904  0.011898    1.310962   
26            0.238368  0.017064    0.401254  1.683336  0.006927    1.272045   
28            0.238368  0.023064    0.456464  1.914955  0.011020    1.401255   
44            0.238368  0.015865    0.404762  1.698053  0.006522    1.279541   
55            0.174110  0.017064    0.416938  2.394681  0.009938    1.416470   
56            0.238368  0.017064    0.435374  1.826477  0.007722    1.348914   
58            0.238368  0.015731    0.443609  1.861024  0.007278    1.368879   

    zhangs_metric  
15       0.474369  
23       0.461536  
26       0.423972  
28       0.503221  
44       0.427860  
55       0.607262  
56       0.470957  
58       0.479672

### 3.b. A partir de `rules`, recupere as regras com lift maior ou igual a 2 e confiança maior ou igual a 0.5. 

Armazene-a na variável `rules_b`.

Chame a função `q3.b.check()` no fim da célula para verificar a resposta.
Você pode pedir dicas com `q3.b.hint()`

In [ ]:
# Resposta da 3.b

# q3.b.hint()

rules_b = rules[(rules['lift'] >= 2) & (rules['confidence'] >= 0.5)]

q3.b.check()
rules_b

### 3.c. Quais produtos você recomendaria colocar próximos se fosse organizar um supermercado físico?

Justifique sua resposta com evidências plausíveis em uma resposta completa.

Essa questão não possui check automático. Mas você pode pedir dica com `q3.c.hint()`. 

**RESPOSTA 3.c.:** {{ESCREVA AQUI}}

#### Os produtos que eu recomendaria colocar próximos são os que possuem maior lift, pois isso significa que eles são comprados juntos com mais frequência. Por exemplo, água mineral seria importante estar próxima de leite e beef, pois o lift entre eles é > que 2.0, o que significa que eles são comprados juntos com frequência. Aliás água mineral é um parametro bem interessante para se analisar, pois ela possui um lift alto com vários produtos, o que significa que ela é um produto que é comprado com frequência e que pode ser na entrada e saída do supermercado.

## 4. Outras questões
Essa questão não possui correção automática.

### 4.a. Liste 5 cenários hipotéticos que você vê utilidade de se analisar regras de associação (pense em algum problema real, algum problema de pesquisa ou desenvolvimento que já atuou ou atua,...)

**RESPOSTA 4.a.:**

1. Análise de comportamento do cliente: : Uma empresa de serviços financeiros está interessada em entender melhor o comportamento de seus clientes em relação aos produtos financeiros oferecidos.
2. Análise de redes sociais: Em plataformas de redes sociais, a análise de regras de associação pode ser usada para identificar conexões e padrões entre os usuários.
3. Sistemas de recomendação de filmes e músicas: Serviços de streaming podem utilizar a análise de regras de associação para identificar associações entre filmes, programas de TV, músicas ou artistas.
4. Manutenção preditiva: Em ambientes industriais, a análise de regras de associação pode ser aplicada para identificar associações entre falhas de equipamentos
5. Medicina e saúde: Em um estudo clínico, pesquisadores podem analisar regras de associação entre sintomas, histórico médico e resultados de testes para identificar padrões que possam indicar condições médicas específicas.

### 4.b. Considerando os resultados das regras, faça uma análise da regra (pancakes, mineral water)-> (spaghetti) em termos das métricas de "antecedent support",	"consequent support", "support",	"confidence" e	"lift". O objetivo é verificar se você entendeu os conceitos dessas métricas. Por exemplo, explique o que significa para essa regra um valor de "antecedent support" de 0.033729. Faça isso para todas as métricas.

In [ ]:
#RESPOSTA 4.b.:

rules_PMS = rules[(rules['antecedents'] == {'pancakes', 'mineral water'}) & (rules['consequents'] == {'spaghetti'})]
rules_PMS

### RESPOSTA 4.b.:

#### Antecedent support: 0.033729 -> O valor 0.033729 indica a frequência relativa do antecedente da regra no conjunto de transações. Isso significa que o antecedente aparece em cerca de 3,37% das transações analisadas.
#### Consequent support: 0.17411 -> O valor 0.17411 indica a frequência relativa do consequente da regra no conjunto de transações. Isso significa que o consequente aparece em cerca de 17,41% das transações analisadas.
#### Support: 0.011465 -> O valor 0.011465 indica a frequência relativa do conjunto de itens (antecedente e consequente) da regra no conjunto de transações. Isso significa que o conjunto de itens aparece em cerca de 1,15% das transações analisadas.
#### Confidence: 0.339921 -> O valor 0.339921 indica a probabilidade condicional de que o consequente ocorra, dado que o antecedente ocorreu. Isso significa que, quando o antecedente da regra é observado, há uma probabilidade de aproximadamente 33,99% de que o consequente também ocorra.
#### Lift: 1.9512 ->  O valor 1.9512 indica o incremento (ou aumento) observado na ocorrência conjunta do antecedente e do consequente em relação à ocorrência esperada se eles fossem independentes um do outro. Um valor de lift acima de 1 indica uma associação positiva entre os itens, neste caso, cerca de 1,95 vezes mais provável de ocorrerem juntos do que se fossem independentes.
#### Leverage: 0.005572 -> O valor 0.005572 indica o incremento absoluto na ocorrência conjunta do antecedente e do consequente em relação à ocorrência esperada se eles fossem independentes. Isso fornece uma medida da associação dos itens, sendo um valor positivo quando a ocorrência conjunta é maior do que a esperada.
#### Conviction: 1.2145 ->  O valor 1.2145 indica a força da dependência entre o antecedente e o consequente, levando em consideração a probabilidade de que a regra seja uma falsa descoberta. Quanto maior o valor de convicção, mais forte é a dependência entre os itens.
#### Zhangs metric: 0.011465 -> ?





### 4.c. Utilize a função auxiliar abaixo para calcular conjuntos de itens frequentes (utilize suporte mínimo no intervalo [0.05, 0.3]) variando os tamanhos. Qual padrão você observa?
Qual variável você acredita que afeta o desempenho desse método? Experimente e discorra sobre quando usar esse método.

In [ ]:
# Gera uma matriz 2D de strings aleatórias.
#
# Parâmetro dv (different values): Quantidade de strings aleatórias únicas que a matriz irá conter.
# Parâmetro vpi (values per instance): Quantidade de strings aleatórias que cada instância terá.
# Parâmetro n: Quantidade de instâncias que será retornada.
def generate_fictional_data(dv=50, vpi=10, n=25000):
    instances = [''.join(random.choices(string.ascii_uppercase + string.digits, k=20)) for _i in range(dv)]
    fictional_data = [random.choices(instances, k=vpi) for _i in range(n)]
    return fictional_data

**RESPOSTA 4.c.:** {{Não sei se consegui fazer certo, porque o support ficou muito parecido (0.18}, mas a ideia é que quanto menor o suporte passado na função a priori, mais itens frequentes são gerados, porém, demorando um tempo absurdo de execução, sendo muito caro computacionalmente. Com min_support menor que 0.03, consegui ver essa diferença de tempos de execução e custos computacionais.}

In [ ]:
#RESPOSTA 4.
fictional = generate_fictional_data()
fictional_array = []

In [ ]:
for i in fictional:
    fictional_array.append(i)
fictional_array


In [ ]:
fictional_t = TransactionEncoder()
fictional_t_array = fictional_t.fit(fictional_array).transform(fictional_array)
fictional_t_array

In [ ]:
fictional_t_df = pd.DataFrame(fictional_t_array, columns=fictional_t.columns_)
fictional_t_df

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.05, use_colnames=True)
frequent_item

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.1, use_colnames=True)
frequent_item

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.15, use_colnames=True)
frequent_item

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.2, use_colnames=True)
frequent_item

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.25, use_colnames=True)
frequent_item

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.3, use_colnames=True)
frequent_item

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.03, use_colnames=True)
frequent_item # demora 2 segundos

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.02, use_colnames=True)
frequent_item # demora 3 segundos

In [ ]:
frequent_item = apriori(fictional_t_df, min_support=0.005, use_colnames=True)
frequent_item # demora 1 minuto :)